File: python-algorithms/src/longest_substring_without_repeating_characters.py
class LongestSubstringWithoutRepeatingCharacters:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()

        a = 0
        b = 0
        longest = 0

        for c in s:
            if c in chars:
                while c in chars:
                    c_aux = s[a]
                    chars.remove(c_aux)
                    a += 1  # moving pointer A

            chars.add(c)
            b += 1  # moving pointer B

            delta = b - a
            longest = max(longest, delta)

        return longest              

File: java-algorithms/src/main/java/com/omsprog/algorithms/lc/LongestSubstringWithoutRepeatingCharacters.java
package com.omsprog.algorithms.lc;

import java.util.HashSet;
import java.util.Set;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> chars = new HashSet<>();

        int a = 0;
        int b = 0;
        int longest = 0;

        for(char c : s.toCharArray()) {
            if(chars.contains(c))
                while(chars.contains(c)) {
                    char cAux = s.charAt(a);
                    chars.remove(cAux);
                    ++a; // moving pointer A
                }

            chars.add(c);
            ++b; // moving pointer B

            int delta = b - a;
            longest = Math.max(longest, delta);
        }

        return longest;
    }
}
